### DISCUSSION TITLE
Moving to .NET5 and .NET6

### DISCUSSION THREAD

**Discussion by amerkoleci** [amerkoleci/Vortice.Windows#196] [General]
Hi guys,
.NET 6 is arround the corner and I plan to move from .NET Standard 2.0 to .NET5 e .NET6, to take advantage of new language features and improvements (MemoryMarshal, Span) etc.

If you find any trouble feel free to suggest.

@SuRGeoNix, @h1cks, @andrew-boyarshin

**Comment by SuRGeoNix** [comment]
I think by removing .NET standard 2.0 support will brake the compatibility with .NET Framework. Correct if I'm wrong here!

[Related source](https://github.com/dotnet/runtime/discussions/40304)

**Comment by h1cks** [comment]
I'm supportive of the move to .NET 6.0 once its mainstream and part of the releases for win 10/11.

I shot myself in the foot when I moved to .NET 5.0 before it became a full release on win 10 (rather than a download from MS).  The interaction with triggering the .NET 5.0 download broke because MS moved it's on download page.

I'm building in .NET 5.0 at the moment, but still using Sharpdx (until I get the FMV working).  I agree with @SuRGeoNix though, my tools are written WPF .NET framework, which might pose an issue for me if .NET Framework breaks (when I migrate my tools to Vortice).

Edit: Windows Forms also supports .NET Core 3.1 and .NET 5.0 which is my real concern so there maybe no issue for me here.

**Reply by amerkoleci** [reply]
WPF and Windows Forms can already be used with .NET5-windows and .NET6-windows, so thats no more an issue.
I think in meanwhile I can move to .NET 5 and wait for .NET 6 once release ofc.

**Reply by h1cks** [reply]
I have been running my tools with .NET 5.0, I'm not completely familiar with the underlying .NET Standard 2.0 compatibility issues of retiring.  From what I understand, .NET Standard 2.0 seems to be caught in the middle of .NET Core/.NET.  Is this assumption correct?

**Reply by amerkoleci** [reply]
> I have been running my tools with .NET 5.0, I'm not completely familiar with the underlying .NET Standard 2.0 compatibility issues of retiring. From what I understand, .NET Standard 2.0 seems to be caught in the middle of .NET Core/.NET. Is this assumption correct?

Yes, but I don't plan to support for so long legacy .NET framework, .NET5 is out

**Reply by h1cks** [reply]
I see no issues with the migration at the end of the day.  The library is relatively new, and will garner more support as compatibility in Sharpdx falls off and associated projects need the new and more modern frameworks.  Your project is very important in terms of the current landscape of DX/C#

**Comment by andrew-boyarshin** [comment]
I'm against this move. .NET Standard (+ .NET Framework) is still too popular to be dismissed. SharpGen even supports 1.1+ (will become 1.3+ soon, though). I understand that you don't want to add too many TFMs (like `SharpGen.Runtime` does, to support the widest range of implementations possible, but use the most performant implementation on each one). There is only one difference between the code generated by SharpGen for .NET versions: .NET 5+ function pointers. I propose adding `net5.0` TFM to all projects, but keeping `netstandard2.0`, thanks to `System.Memory` NuGet package. Do you really need to require .NET Core? Popular and important projects like Avalonia still support .NET Standard 2.0, and, from what I gather, there are no immediate plans for that to change.

**Reply by andrew-boyarshin** [reply]
Note, that personally, I use latest .NET 7.0 daily builds. But I know a lot of enterprises remain at .NET Core 3.1 at best, at .NET Framework 4.5 at the worst. Intentional deprecation without proper cause reflects badly on .NET ecosystem.

**Reply by amerkoleci** [reply]
I see, also wave engine, which is using vortice windows is targeting netstandard 2.0, I'll add net5 TFM and keep netstandard2.0 as you suggest.

Thanks

**Reply by andrew-boyarshin** [reply]
Note that a lot of C# 8.0+ features don't really depend on .NET Runtime. You can use a lot! Here's somewhat out of date summary by @Sergio0694: [Medium link](https://sergiopedri.medium.com/enabling-and-using-c-9-features-on-older-and-unsupported-runtimes-ce384d8debb), without C# 10 features or all the workarounds for some features (links to some NuGet packages with ready-made compat shims).

**Reply by amerkoleci** [reply]
This was the code I took from TerraFX that didn't work under .NET Standard 2.0, not trivial bu:

https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.DirectX/UnsafeUtilities.cs#L139

**Reply by andrew-boyarshin** [reply]
@amerkoleci I see a lot of duplication between the code in this class and the existing facilities provided by `Unsafe` class and `MemoryHelpers` from _SharpGen.Runtime_.

**Reply by amerkoleci** [reply]
I saw that and remove all the unused stuff from UnsafeUtilties, https://github.com/amerkoleci/Vortice.Windows/commit/7b802b3286f078598c89ca2981a75d902b6f76c7 adds .NET 5 as TFM supported.

**Reply by SuRGeoNix** [reply]
@amerkoleci Ideally, it is too early to drop .net standard 2.0 => .net framework support and I don't really think you have any reason to do it. The only reason will be if you want to use new features from .NET 5/6 which will lead you to write different code for each framework but I don't believe it will be that much that worth dropping support now.

**Reply by amerkoleci** [reply]
@andrew-boyarshin Any thoughts why CI build fails for Vortice.DirectInput? https://github.com/amerkoleci/Vortice.Windows/runs/3882555368?check_suite_focus=true

**Reply by andrew-boyarshin** [reply]
@amerkoleci SharpGen bug in `Sdk.targets`. Proposed workaround: do not depend on multi-TFM projects (like DXGI) until the fix is made. Or switch to `PackageReference`, but that's harder.

**Reply by amerkoleci** [reply]
Do you plan to fix this bug? Otherwise can you send me workaround pr please?

**Reply by andrew-boyarshin** [reply]
I think I can make a hotfix, yeah. ETA: 24 hours.

**Reply by amerkoleci** [reply]
Don't worry, no rush! Thank you

**Reply by amerkoleci** [reply]
> I think I can make a hotfix, yeah. ETA: 24 hours.

Any news here dude?
